From a4c177cc8c22eda5996343466ba3c87e286f2093 Mon Sep 17 00:00:00 2001
From: Dhruva Gole <d-gole@ti.com>
Date: Wed, 5 Feb 2025 14:27:53 -0600
Subject: [PATCH 02/68] TI: HACK: arm64: dts: ti: Configure am62a
 pinctrl-single for wake-up interrupts

Configure am62a pinctrl-single for wake-up interrupts.

Based on the patch for am625 with changes to the length of the register
set and the interrupt ID. The length has been updated to match
end address - start address for Main PADCFG registers. The interrupt ID
has been updated to next available line than what was used on AM625.

Marked as HACK because physically none of the wakeup source events are
connected to GIC. Eventually we may want to set up a chained interrupt
handler in the ti-msgmgr for pinctrl-single to use.

[k-willis@ti.com] change to append compatible property with pinctrl-single
to fix DT warning. Change to update reg length. Squashed commit
3c2a5ceb94df ("arm64: dts: ti: k3-am62a7-sk: Fix pinctrl properties") and
relevant part of commit eb3b9e22be86 ("arm64: dts: ti: k3-am62*:
Append pinctrl-single compatible").

Signed-off-by: Vibhore Vardhan <vibhore@ti.com>
Signed-off-by: Dhruva Gole <d-gole@ti.com>
Signed-off-by: Devarsh Thakkar <devarsht@ti.com>
Reported-by: kernel test robot <lkp@intel.com>
Reviewed-by: Dhruva Gole <d-gole@ti.com>
Signed-off-by: Kendall Willis <k-willis@ti.com>
---
 arch/arm64/boot/dts/ti/k3-am62a-main.dtsi | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/ti/k3-am62a-main.dtsi b/arch/arm64/boot/dts/ti/k3-am62a-main.dtsi
index 174bc03d3037..0b47fdf487ac 100644
--- a/arch/arm64/boot/dts/ti/k3-am62a-main.dtsi
+++ b/arch/arm64/boot/dts/ti/k3-am62a-main.dtsi
@@ -257,11 +257,24 @@ secure_proxy_sa3: mailbox@43600000 {
 	};
 
 	main_pmx0: pinctrl@f4000 {
-		compatible = "pinctrl-single";
-		reg = <0x00 0xf4000 0x00 0x2ac>;
+		compatible = "ti,am654-padconf", "pinctrl-single";
+		reg = <0x00 0xf4000 0x00 0x25c>;
 		#pinctrl-cells = <1>;
 		pinctrl-single,register-width = <32>;
 		pinctrl-single,function-mask = <0xffffffff>;
+		/*
+		 * pinctrl IP DOES NOT give any functional IRQs when a
+		 * system is in active state. This IRQ is only a dummy IRQ
+		 * that is being used to trick Linux into thinking that
+		 * GIC can potentially recieve an interrupt from this IP.
+		 * This helps us setup the IO daisychain wakeups for deep
+		 * sleep via chained wake IRQs.
+		 * Please feel free to assign a different number here as
+		 * long as it is unused if 99 conflicts with another use case.
+		 */
+		interrupts = <GIC_SPI 99 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-controller;
+		#interrupt-cells = <1>;
 	};
 
 	main_esm: esm@420000 {
-- 
2.47.2

